sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -464,6 +465,24 @@ object OptimizeIn extends Rule[LogicalPlan] {
 }
 
 /**
+  * Convert an expression into its conjunctive normal form (CNF), i.e. AND of ORs.
+  * For example, a && b || c is normalized to (a || c) && (b || c) by this method.
+  *
+  * Refer to https://en.wikipedia.org/wiki/Conjunctive_normal_form for more information
+  */
+object CNFNormalization extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case q: LogicalPlan => q transformExpressionsUp {
+      // reverse Or with its And child to eliminate (And under Or) occurrence
+      case Or(And(innerLhs, innerRhs), rhs) =>
+        And(Or(innerLhs, rhs), Or(innerRhs, rhs))
+      case Or(lhs, And(innerLhs, innerRhs)) =>
+        And(Or(lhs, innerLhs), Or(lhs, innerRhs))
+    }
+  }
+}
```

Comment:
We should probably include/move those `Not` push-down rules added in PR #5700 defined in `BooleanSimplification` here, since they are essential for CNF normalization.